// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

edition = "2023";

package brain.evolution;

import "evolution/lib/population_client_spec.proto";
import "evolution/lib/population_server_spec.proto";

message PopulationClientManagerSpec {
  repeated PopulationClientManagerElement elements = 1;
}

message PopulationClientManagerElement {
  // The client spec for constructing this client element. Required.
  PopulationClientSpec client_spec = 1;

  // The server spec used to construct a non-remote server. Required.
  PopulationServerSpec local_server_spec = 2;

  // The probability of choosing this client (higher = more likely). Required.
  double weight = 3 [default = 1.0];

  // RPC throttling specification. The client will wait for at least this
  // amount of time between requests through this client. (In practice, it can
  // be larger due to server response times or worker computation times).
  double target_rpc_period_secs = 5 [default = 0.0];
}
